xray/src/bin/merge_xray_quadtrees.rs

```diff
@@ -20,8 +24,156 @@ struct CommandlineArguments {
+    for input_directory in input_directories {
+        copy_images(input_directory, output_directory);
+    }
+}
+
+fn read_metadata_from_directory(directory: &Path) -> Vec<Meta> {
+    globwalk::GlobWalkerBuilder::new(directory, "meta*.pb")
+        .build()
+        .expect("Failed to build GlobWalker")
+        .filter_map(Result::ok)
+        .map(|dir_entry| {
+            Meta::from_disk(dir_entry.path())
+                .expect(&format!("Failed to load meta from {:?}.", dir_entry.path()))
+        })
+        .collect()
+}
+
+fn read_metadata_from_directories(directories: &[PathBuf]) -> Vec<Meta> {
+    directories
+        .iter()
+        .map(|directory| read_metadata_from_directory(&directory))
+        .flatten()
+        .collect()
+}
+
+fn get_root_node(meta: &Meta) -> Option<NodeId> {
+    meta.nodes.iter().copied().min_by_key(|node| node.level())
+}
+
+fn get_root_nodes(meta: &[Meta]) -> Vec<NodeId> {
+    let root_nodes: Vec<NodeId> = meta.iter().filter_map(get_root_node).collect();
+    if root_nodes.len() != meta.len() {
+        println!(
+            "Skipped {} empty subquadtrees.",
+            meta.len() - root_nodes.len()
+        );
+    }
+    root_nodes
+}
+
+struct MergedMetadata {
+    root_nodes: FnvHashSet<NodeId>,
+    level: u8,
+    merged_meta: Meta,
+}
+
+// Checks wheter all the elements in the iterator have the same value and returns it.
+// There must be at least one element in the iterator.
+fn check_all_the_same<I, V>(mut iterator: I) -> Option<V>
+where
+    I: Iterator<Item = V> + Clone,
+    V: std::cmp::PartialEq,
+{
+    if iterator.clone().all_equal() {
+        Some(iterator.next().expect("Iterator cannot be empty."))
+    } else {
+        None
+    }
+}
+
+fn validate_and_merge_metadata(metadata: &[Meta]) -> MergedMetadata {
+    assert!(!metadata.is_empty(), "No subquadtrees meta files found.");
+    let root_nodes_vec = get_root_nodes(metadata);
+    let root_nodes: FnvHashSet<NodeId> = root_nodes_vec.iter().copied().collect();
+    assert_eq!(
+        root_nodes.len(),
+        root_nodes_vec.len(),
+        "Not all roots are unique."
+    );
+    let level = check_all_the_same(root_nodes.iter().map(|node| node.level()))
+        .expect("Not all roots have the same level.");
+    let deepest_level = check_all_the_same(metadata.iter().map(|meta| meta.deepest_level))
+        .expect("Not all meta files have the same deepest level.") as u8;
+    let tile_size = check_all_the_same(metadata.iter().map(|meta| meta.tile_size))
+        .expect("Not all meta files have the same tile size.");
+    // TODO: check whether all root nodes have the same bound_rect.
+    let bounding_rect = metadata[0].bounding_rect.clone();
+
+    let mut nodes = FnvHashSet::default();
+    for meta in metadata {
+        nodes.extend(&meta.nodes);
+    }
+
+    MergedMetadata {
+        root_nodes,
+        level,
+        merged_meta: Meta {
```

Comment:
Nit: I think we can remove the `merged_` prefix here or rename it to `root_meta`... What do you think?

Reply:
:+1: sounds good